File: scripts/swap-pool/swapAmmV2.ts
import * as idl from "../../target/idl/swap_pool.json";

import { SwapPool } from "../../target/types/swap_pool";
import {
  AddressLookupTableAccount,
  clusterApiUrl,
  Connection,
  LAMPORTS_PER_SOL,
  PublicKey,
  SystemProgram,
  Transaction,
} from "@solana/web3.js";
import {
  getAssociatedTokenAddressSync,
  NATIVE_MINT,
  getOrCreateAssociatedTokenAccount,
  TOKEN_PROGRAM_ID,
  getAccount,
} from "@solana/spl-token";

import { initSdk } from "../config";
import { payer } from "../wallet";
import { BN } from "bn.js";
import { addAddressesToTable, createLookupTable } from "../lookupTable";
import { finalizeTransaction, wrappedSOLInstruction } from "../utils";
import { TokenInfo } from "../type";
import { SwapPoolProgram } from "../../sdk/swap-pool-program";
import { tokens } from "../token";

const connection = new Connection(clusterApiUrl("devnet"), "confirmed");

const swapPoolProgram = new SwapPoolProgram(idl as SwapPool, connection);

export async function swapAmmV2(
  tokensInfo: TokenInfo[],
  inputMint: PublicKey,
  amountIn: number,
  minimumAmountOut: number
) {
  const raydium = await initSdk();
  const tx = new Transaction();
  const lookupTables: AddressLookupTableAccount[] = [];

  const lookupTableAddress = new PublicKey(
    "FFnNLnvHcUxuYCvBAjtj33Yt9xmMAsHZgwnRB6WwJhqS"
  );
  const lookupTable = (
    await connection.getAddressLookupTable(lookupTableAddress)
  ).value;

  for (const tokenInfo of tokensInfo) {
    const { instructions, lookupTable } = await processTokenSwap(
      connection,
      raydium,
      swapPoolProgram,
      payer,
      tokenInfo,
      inputMint,
      amountIn,
      minimumAmountOut
    );

    tx.instructions.push(...instructions);

    // if (lookupTable) {
    //   const lut = (await connection.getAddressLookupTable(lookupTable)).value;
    //   if (lut) {
    //     lookupTables.push(lut);
    //   }
    // }
  }

  await finalizeTransaction(connection, payer, tx, [lookupTable]);
}

async function processTokenSwap(
  connection: Connection,
  raydium: any,
  swapPoolProgram: SwapPoolProgram,
  payer: any,
  tokenInfo: TokenInfo,
  inputMint: PublicKey,
  amountIn: number,
  minimumAmountOut: number
): Promise<{ instructions: any[]; lookupTable: PublicKey | null }> {
  const txInstructions: any[] = [];
  let lookupTable: PublicKey | null = null;

  try {
    const data = await raydium.liquidity.getPoolInfoFromRpc({
      poolId: tokenInfo.ammId,
    });
    if (!data || !data.poolKeys) {
      return;
    }

    const poolKeys = data.poolKeys;
    const baseIn = inputMint.toString() === poolKeys.mintA.address;

    const [mintIn, mintOut] = baseIn
      ? [poolKeys.mintA.address, poolKeys.mintB.address]
      : [poolKeys.mintB.address, poolKeys.mintA.address];

    const inputTokenAccount = getAssociatedTokenAddressSync(
      new PublicKey(mintIn),
      payer.publicKey,
      false
    );

    console.log("poolKeys: ", poolKeys);

    const outputTokenAccount = await getOrCreateAssociatedTokenAccount(
      connection,
      payer,
      new PublicKey(mintOut),
      swapPoolProgram.configPDA,
      true
    );

    const vaultTokenInfo = await getAccount(
      connection,
      outputTokenAccount.address
    );
    if (inputMint.equals(NATIVE_MINT)) {
      const wrappedSolIx = await wrappedSOLInstruction(
        payer.publicKey,
        amountIn
      );
      txInstructions.push(...wrappedSolIx);
    }

    const swapAmmV2Ix = await swapPoolProgram.program.methods
      .swapAmmV2(new BN(amountIn), new BN(minimumAmountOut))
      .accountsPartial({
        userSourceOwner: payer.publicKey,
        amm: new PublicKey(tokenInfo.ammId),
        ammAuthority: new PublicKey(poolKeys.authority),
        ammOpenOrders: new PublicKey(poolKeys.openOrders),
        ammCoinVault: new PublicKey(poolKeys.vault.A),
        ammPcVault: new PublicKey(poolKeys.vault.B),
        marketProgram: new PublicKey(poolKeys.marketProgramId),
        market: new PublicKey(poolKeys.marketId),
        marketBids: new PublicKey(poolKeys.marketBids),
        marketAsks: new PublicKey(poolKeys.marketAsks),
        marketEventQueue: new PublicKey(poolKeys.marketEventQueue),
        marketCoinVault: new PublicKey(poolKeys.marketBaseVault),
        marketPcVault: new PublicKey(poolKeys.marketQuoteVault),
        marketVaultSigner: new PublicKey(poolKeys.marketAuthority),
        ammProgram: new PublicKey(poolKeys.programId),
        userTokenSource: inputTokenAccount,
        vaultTokenAccount: outputTokenAccount.address,
      })
      .instruction();

    const accountKeys = [
      new PublicKey(tokenInfo.ammId),
      new PublicKey(poolKeys.authority),
      new PublicKey(poolKeys.openOrders),
      new PublicKey(poolKeys.vault.A),
      new PublicKey(poolKeys.vault.B),
      new PublicKey(poolKeys.marketProgramId),
      new PublicKey(poolKeys.marketId),
      new PublicKey(poolKeys.marketBids),
      new PublicKey(poolKeys.marketAsks),
      new PublicKey(poolKeys.marketEventQueue),
      new PublicKey(poolKeys.marketBaseVault),
      new PublicKey(poolKeys.marketQuoteVault),
      new PublicKey(poolKeys.marketAuthority),
      new PublicKey(poolKeys.programId),
      SystemProgram.programId,
      swapPoolProgram.program.programId,
      TOKEN_PROGRAM_ID,
    ];

    if (tokenInfo.lut) {
      lookupTable = tokenInfo.lut;
    } else {
      lookupTable = await createLookupTable(connection, payer);
      await addAddressesToTable(connection, payer, lookupTable, accountKeys);
      console.log(`Lookup Table Address: ${lookupTable}`);
    }

    txInstructions.push(swapAmmV2Ix);
  } catch (error) {
    console.error(
      `Error processing token swap for AMM ID: ${tokenInfo.ammId}`,
      error
    );
  }

  return { instructions: txInstructions, lookupTable };
}

const amountIn = 0.1 * LAMPORTS_PER_SOL;
const minAmountOut = 0;

swapAmmV2(tokens.slice(0, 4), NATIVE_MINT, amountIn, minAmountOut);
swapAmmV2(tokens.slice(4, 8), NATIVE_MINT, amountIn, minAmountOut);
swapAmmV2(tokens.slice(8, 12), NATIVE_MINT, amountIn, minAmountOut);
swapAmmV2(tokens.slice(12, 16), NATIVE_MINT, amountIn, minAmountOut);

File: programs/swap-pool/src/instructions/swap_amm_v2.rs
use anchor_lang::prelude::*;
use anchor_spl::token::{Token, TokenAccount};

use raydium_amm_cpi::*;

use crate::ConfigAccount;

#[derive(Accounts)]
pub struct SwapAMMV2<'info> {
    /// CHECK: Safe. user owner Account
    #[account(mut)]
    pub user_source_owner: Signer<'info>,
    #[account(
        mut,
        seeds = [b"config".as_ref()],
        bump
    )]
    pub config_account: Account<'info, ConfigAccount>,
    /// CHECK: Safe. amm Account
    #[account(mut)]
    pub amm: AccountInfo<'info>,
    /// CHECK: Safe. Amm authority Account
    #[account(mut)]
    pub amm_authority: AccountInfo<'info>,
    /// CHECK: Safe. amm open_orders Account
    #[account(mut)]
    pub amm_open_orders: AccountInfo<'info>,
    /// CHECK: Safe. amm_coin_vault Amm Account to swap FROM or To,
    #[account(mut)]
    pub amm_coin_vault: AccountInfo<'info>,
    /// CHECK: Safe. amm_pc_vault Amm Account to swap FROM or To,
    #[account(mut)]
    pub amm_pc_vault: AccountInfo<'info>,
    /// CHECK: Safe.OpenBook program id
    pub market_program: AccountInfo<'info>,
    /// CHECK: Safe. OpenBook market Account. OpenBook program is the owner.
    #[account(mut)]
    pub market: AccountInfo<'info>,
    /// CHECK: Safe. bids Account
    #[account(mut)]
    pub market_bids: AccountInfo<'info>,
    /// CHECK: Safe. asks Account
    #[account(mut)]
    pub market_asks: AccountInfo<'info>,
    /// CHECK: Safe. event_q Account
    #[account(mut)]
    pub market_event_queue: AccountInfo<'info>,
    /// CHECK: Safe. coin_vault Account
    #[account(mut)]
    pub market_coin_vault: AccountInfo<'info>,
    /// CHECK: Safe. pc_vault Account
    #[account(mut)]
    pub market_pc_vault: AccountInfo<'info>,
    /// CHECK: Safe. vault_signer Account
    #[account(mut)]
    pub market_vault_signer: AccountInfo<'info>,
    /// CHECK: Safe. user source token Account. user Account to swap from.
    #[account(mut)]
    pub user_token_source: AccountInfo<'info>,
    /// CHECK: Safe. user destination token Account. user Account to swap to.
    #[account(
        mut,
        constraint = vault_token_account.owner == config_account.key()
    )]
    pub vault_token_account: Account<'info, TokenAccount>,
    /// CHECK: Safe. The spl token program
    pub token_program: Program<'info, Token>,
    /// CHECK: Safe. amm_program
    pub amm_program: AccountInfo<'info>,
}

pub fn swap_amm_v2(ctx: Context<SwapAMMV2>, amount_in: u64, minimum_amount_out: u64) -> Result<()> {
    let cpi_accounts = SwapBaseIn {
        amm: ctx.accounts.amm.clone(),
        amm_authority: ctx.accounts.amm_authority.clone(),
        amm_open_orders: ctx.accounts.amm_open_orders.clone(),
        amm_coin_vault: ctx.accounts.amm_coin_vault.clone(),
        amm_pc_vault: ctx.accounts.amm_pc_vault.clone(),
        market_program: ctx.accounts.market_program.clone(),
        market: ctx.accounts.market.clone(),
        market_bids: ctx.accounts.market_bids.clone(),
        market_asks: ctx.accounts.market_asks.clone(),
        market_event_queue: ctx.accounts.market_event_queue.clone(),
        market_coin_vault: ctx.accounts.market_coin_vault.clone(),
        market_pc_vault: ctx.accounts.market_pc_vault.clone(),
        market_vault_signer: ctx.accounts.market_vault_signer.clone(),
        user_token_source: ctx.accounts.user_token_source.clone(),
        user_token_destination: ctx.accounts.vault_token_account.to_account_info(),
        user_source_owner: ctx.accounts.user_source_owner.clone(),
        token_program: ctx.accounts.token_program.clone(),
    };

    let cpi_context = CpiContext::new(ctx.accounts.amm_program.to_account_info(), cpi_accounts);
    raydium_amm_cpi::instructions::swap_base_in(cpi_context, amount_in, minimum_amount_out)?;

    Ok(())
}
